### ISSUE TITLE
Sessions are invalidated when idle timeout is set altough there is active comm between client and server

### ISSUE THREAD

**Issue by kovadam69** [hazelcast/hazelcast-wm#73]
Hi!

I run into a problem, where I used WebFilter to put sessions into a cluster map. I set 3600 seconds (1 hour) idle time on the map. Everything was working correctly until I discovered, that despite the client was communicating continuously with the server, the hazelcast sessions were removed after 1 hour from the session map, making a migration from one node to another impossible (I use sticky session, but when a node fails (or I stop it) it forces the user to reauthenticate since the session information is already gone from the session map).

During testing and monitoring the map entries with hazelcast management center I discovered, that after the first session initialization and login the session is not accessed any more in the cluster map (no get/put operations), so after 1 hour (in my test 3 minutes) the session was removed from the session map because of idle time expired. I checked the WebFilter code, and discovered, that actually it uses some kind of internal cache (local cache) to access hazelcast sessions, so normally with sticky session the session on the cluster map is never accessed again, so it can expire.

Extension: my app does not really store huge informations on the session attribute map, so this is why it is not written and read continuously, the session attribute is only storing basic spring authentication data, and such things.

Now I managed to overcame the problem to add a custom filter, which simply increases a request counter in the session attribute map at every request and so I managed to trigger put operation on the cluster session map to update the session cluster-wide, which reset the idle timer, so the session is not getting evicted from the cluster map.

However would be good to get it working without such a hack, so WebFilter should also access the cluster session from the cluster instead of getting it from a local cache, or at least notify somehow the cluster map, that the entry should reset the idle timer. Or to provide some kind of init parameter to trigger this functionality.

**Comment by kovadam69** [comment]
I also discovered that if I turn off idle timeout on the map, than sessions get stuck in the map. I had cluster members having over 70.000 sessions in the session map, most of them should have already gone.
I also encountered issue #72 where multiple 10.000 of session were cached in the WebFilter's sessionMap attribute without having a pair in the originalSessionMap.
However I think the idle timeout solves the problem, because these "false" sessions are then evicted from the session map, and also in this case the WebFilter's destroySession is called which removes these entries from the local cache.

**Comment by nfrankel** [comment]
Hello @kovadam69 

Thanks for using Hazelcast!

Regarding your question, I'm not sure accessing the cluster every time is a good idea, because of the decreased performance. Getting local data is always a safer bet.

**Comment by kovadam69** [comment]
That's true, however the cluster should be notified somehow, that the session is "accessed", because without it, it will evict the session from the cluster map, and make the transition from one node to another impossible.
Now actually I have to change a session attribute for every request, to get the cluster notified, which is at the end the same, as accessing the session from the cluster at the first place. Only advantage is, that I have deferred-write enabled, so it should write all the changes only at the end of the request back to the cluster.

**Comment by enozcan** [comment]
Hi @kovadam69 

Can you please share the configuration of the filter? Also, are you using Hazelcast in Client mode or P2P mode? Regarding to that:
>  I also discovered that if I turn off idle timeout on the map, than sessions get stuck in the map

From the documentation:

> Note that, in the client/server mode of session replication, session-ttl-seconds configuration does not have any effect. The reason is that the filter based session replication uses IMap and a Hazelcast client cannot change the configuration of a distributed map. Instead, you should configure the max-idle-seconds element in your hazelcast.xml on the server side.

**Comment by kovadam69** [comment]
Hi @enozcan 

Here is my web.xml filter configuration. The session-ttl-seconds I do not set neither in the web.xml nor in my spring configuration. I do not want to set session-ttl-seconds, what I would like to use is max-idle-seconds, since I do not want sessions stuck in the map for whatever reason they failed to dispose (destroy the session, eg. tomcat crash, app crash, anything), but I also don't want sessions to be freed after a certain time, since there are clients which are connected continuously and if the session expires after 24h for example, they cannot migrate from one node to another for whatever reason they should be able.
web.xml:
```
<filter>
        <filter-name>hazelcast-filter</filter-name>
        <filter-class>com.hazelcast.web.WebFilter</filter-class>

        <!-- use existing Hazelcast instance created in the messaging service -->
        <init-param>
            <param-name>instance-name</param-name>
            <param-value>bazzle-hz-cluster</param-value>
        </init-param>
        <init-param>
            <param-name>shutdown-on-destroy</param-name>
            <param-value>false</param-value>
        </init-param>
        <init-param>
            <param-name>deferred-write</param-name>
            <param-value>true</param-value>
        </init-param>

        <!-- Name of the distributed map storing web session objects -->
        <init-param>
            <param-name>map-name</param-name>
            <param-value>bazzle-sessions</param-value>
        </init-param>
        <init-param>
            <param-name>sticky-session</param-name>
            <param-value>true</param-value>
        </init-param>
        <init-param>
            <param-name>debug</param-name>
            <param-value>true</param-value>
        </init-param>
    </filter>
```
And the spring configuration of the hazelcast cluster:
```
<hz:hazelcast id="hazelcastInstance">
        <hz:config>
            <hz:instance-name>bazzle-hz-cluster</hz:instance-name>
            <hz:group name="bazzle-cluster" password="${HZ_PASSWORD:default}"/>
            <hz:management-center enabled="${HZ_MAN_CENTER_ENABLED:false}" update-interval="${HZ_MAN_CENTER_UPDATEINTERVAL:3}" url="${HZ_MAN_CENTER_URL:}"/>
            <hz:properties>
                <hz:property name="hazelcast.shutdownhook.enabled">false</hz:property>
            </hz:properties>
            <hz:network port="${HZ_PORT:5701}" port-auto-increment="${HZ_PORT_AUTOINC:true}">
                <hz:join>
                    <hz:multicast enabled="${HZ_MULTICAST:true}"
                                  multicast-group="${HZ_MULTICAST_GROUP:224.2.2.3}"
                                  multicast-port="${HZ_MULTICAST_PORT:54327}"
                                  multicast-timeout-seconds="${HZ_MULTICAST_TIMEOUT:2}"
                                  multicast-time-to-live="${HZ_MULTICAST_TTL:32}"/>
                    <hz:tcp-ip enabled="${HZ_TCPIP:false}">
                        <hz:members>${HZ_TCPIP_MEMBERS:localhost}</hz:members>
                    </hz:tcp-ip>
                    <hz:aws enabled="${HZ_AWS:false}"
                            access-key="${HZ_AWS_ACCESSKEY:default}"
                            secret-key="${HZ_AWS_SECRETKEY:default}"
                            connection-timeout-seconds="${HZ_AWS_CONNTIMEOUT:5}"
                            host-header="${HZ_AWS_HOSTHEADER:ec2.amazonaws.com}"
                            region="${HZ_AWS_REGION:us-east-1}"
                            security-group-name="${HZ_AWS_SECUGROUP:}"
                            tag-key="${HZ_AWS_TAGKEY:}"
                            tag-value="${HZ_AWS_TAGVALUE:}"/>
                </hz:join>
            </hz:network>
            <hz:map name="bazzle-sessions"
                    max-idle-seconds="180"/>
            <hz:map name="bazzle-form-cache"
                    max-size="500"
                    max-size-policy="PER_PARTITION"
                    eviction-percentage="20"
                    eviction-policy="LRU"
                    max-idle-seconds="864000"/>
            <hz:map name="bazzle-report-cache"
                    max-size="500"
                    max-size-policy="PER_PARTITION"
                    eviction-percentage="20"
                    eviction-policy="LRU"
                    max-idle-seconds="864000"/>
        </hz:config>
    </hz:hazelcast>
```
As you see I set max-idle-seconds for the testing here to 180 sec (3 mins) and if I just log in to my application, and work like normal for 3 minutes (so there are continuous request between client and server) the session is removed after 3 minutes from the map although the session is still alive. I added to one of my filters following code, which solved the problem by setting a session attribute each time a request comes in making the session "accessed" in the cluster and resetting the idle time:
```
Object o = session.getAttribute("request_count");
long count = 0;
if (!(o instanceof Number))
                {
                    session.setAttribute("request_count", count);
                }
                else
                {
                    count = ((Number) o).longValue();
                }

                session.setAttribute("request_count", ++count);
```
There is no other way to mark the cluster the session object as "used", there is no get/put operation on the session in the cluster and the idle time can expire although there are lot of requests going on between client and server. The tomcat session is also not expired.
But this is a workaround which I had to do, but the nice way would be that the WebFilter notifies the cluster about the fact, that the session is still alive.

**Comment by enozcan** [comment]
Hi @kovadam69 

Thanks for reporting.
The current behavior of the filter requires a session attribute change to notify distributed map for session activity. A possible solution would be just as you come up with, which unfortunately removes the advantage of the local sessions cache.

In case of interest, here are a few details about local cache and the reason of "stuck session" behavior:

Each session is cached locally as well as in distributed map on creation. For each request, if this session exists in the local cache it's fetched locally in order to reduce to number of calls made to distributed map (due to performance issues). The sessions are not updated on the distributed map unless an attribute is changed. When an HttpSession expires it is evicted from the local cache but not from the distributed map. If it's evicted from the distributed map as well there would be inconsistency for such a scenario:

Say you have tomcat1 and tomcat2 deployed and use the same distributed map for session store. A client first hits tomcat1 and the session is created and stored both locally & in distributed map. Then for all the following requests tomcat2 responses. When the session is expired in tomcat1 due to inactivity, it will trigger the listener to evict the session. However, this session is actually active on tomcat2. Thus the session will be in the local cache of tomcat2 but not in the distributed map (evicted via tomcat1 listener). This is inconsistent. In order to prevent that distributed map entries have to be evicted using TTL or max idle time. That's why sessions are stuck when you disable max idle time configuration.

**Comment by kovadam69** [comment]
Hi @enozcan !

Exactly, I totally agree with you. It makes fully sense not to invalidate sessions in the cluster, except the user/developer uses some kind of logout process. For example I call session.invalidate to invalidate session in tomcat and in the cluster when the user logs out. However this might also not invalidate local session cache on different other nodes, since there is no event listener installed on the map to notify the WebFilter that a session from the map has been evicted, so they can also clean up local cache. This might be also a good addon, to install a listener, and if some other node invalidates the session (or the idle time expires) and the session is removed from the cluster map, the WebFilters on all nodes should check if they have the session in the local cache, and remove them. This might solve the problem that sessions stuck also in the local WebFilter cache.
For example you have 2 nodes, and for some reason the client migrates from one node to another (from node#1 to node#2). Now in the local session cache both nodes have the same session stored. Then the user logs out from the application, the session is invalidated and removed from node#2 and the cluster map, however is not removed from the node#1's local cache, which might stuck there forever consuming lot of memory. I experienced 60-70.000 stuck sessions in the WebFilter's sessionMap variable, however there were only 10-12 sessions in the originalSessionMap, means tens of thousands HazelCastSessions do not have any HttpSession so they are never removed. A listener on the cluster map might help on this, whenever a session is removed from the cluster map, the WebFilter get's notified, and can remove associated HazelcastSession and HttpSession from local cache (they should be invalid anyway, so it is safe to do).

Regarding idle time not updating: indeed accessing the cluster is a costly way to do it, no matter it happens in WebFilter, or I do it manually. Of course I could "fine tune it" to store some kind of timestamp (eg. System.currentTimeMillis) in the attribute map, and only update it, when the previous timestamp was null or at least 30 minutes expired from the last call. Since to keep the session on the tomcat alive, I do use a "ping" mechanism between my client and server, so there is always some kind of communication going on (client is a swing app). But this is some kind of special use case in my case.
May be a better approach would be to update only the idle time property on the map item somehow. This could be done asynchronously in the WebFilter, so it won't stop the filter processing, however still notify the cluster map about the session being alive. I don't know if there is such a method, where you can notify the cluster to reset the idle time on a particular map's element. If there is not, may be simply making a "get" operation on the map in a separate Thread would notify the map about the session being used, and won't stuck the filter processing. This would sacrify still some performance, but the user experience won't be affected that much.

Thanks!

**Comment by enozcan** [comment]
Hi @kovadam69 

Thanks for sharing your ideas. 
IMO they make sense and might be included in the filter.

**Comment by enozcan** [comment]
Hi @kovadam69 
A fix concerning the issue is merged into the master with fc0df8232ab38b2554bcd09a5c8eff1bcb0c4432. 
Closing the issue. Feel free to reopen if you have a question about the new feature.
